hudi-utilities/src/main/java/org/apache/hudi/utilities/HoodieDataTableValidator.java

```diff
@@ -0,0 +1,473 @@
+          danglingFilePaths.forEach(entry -> LOG.error("Dangling file: " + entry.toString()));
+          finalResult = false;
+          if (!cfg.ignoreFailed) {
+            throw new HoodieValidationException("Data table validation failed due to dangling files " + danglingFilePaths.size());
+          }
+        }
+
+        // Verify that for every completed commit in active timeline, there are no extra files found apart from what is present in
+        // commit metadata.
+        Map<String, List<String>> instantToFilesMap = RepairUtils.tagInstantsOfBaseAndLogFiles(
+            metaClient.getBasePath(), allDataFilePaths);
+        HoodieActiveTimeline activeTimeline = metaClient.getActiveTimeline();
+        List<HoodieInstant> hoodieInstants = activeTimeline.filterCompletedInstants().getInstants().collect(Collectors.toList());
+
+        List<String> danglingFiles = engineContext.flatMap(hoodieInstants, instant -> {
+          Option<Set<String>> filesFromTimeline = RepairUtils.getBaseAndLogFilePathsFromTimeline(
+              activeTimeline, instant);
+          List<String> baseAndLogFilesFromFs = instantToFilesMap.containsKey(instant.getTimestamp()) ? instantToFilesMap.get(instant.getTimestamp())
+              : Collections.emptyList();
+          if (!baseAndLogFilesFromFs.isEmpty()) {
+            Set<String> danglingInstantFiles = new HashSet<>(baseAndLogFilesFromFs);
+            if (filesFromTimeline.isPresent()) {
+              danglingInstantFiles.removeAll(filesFromTimeline.get());
+            }
+            return new ArrayList<>(danglingInstantFiles).stream();
+          } else {
+            return Stream.empty();
+          }
+        }, hoodieInstants.size()).stream().collect(Collectors.toList());
+
+        if (!danglingFiles.isEmpty()) {
+          LOG.error("Data table validation failed due to extra files found for completed commits " + danglingFiles.size());
+          danglingFiles.forEach(entry -> LOG.error("Dangling file: " + entry.toString()));
+          finalResult = false;
+          if (!cfg.ignoreFailed) {
+            throw new HoodieValidationException("Data table validation failed due to dangling files " + danglingFiles.size());
+          }
+        }
+      }
+    } catch (Exception e) {
+      LOG.error("Data table validation failed due to " + e.getMessage(), e);
+      if (!cfg.ignoreFailed) {
+        throw new HoodieValidationException("Data table validation failed due to " + e.getMessage(), e);
+      }
+    }
+
+    if (finalResult) {
+      LOG.info("Data table validation succeeded.");
+    } else {
+      LOG.warn("Data table validation failed.");
+    }
+  }
+
+  /**
+   * Compare the listing partitions result between metadata table and fileSystem.
+   */
+  private List<Path> getAllFiles(HoodieSparkEngineContext engineContext, String basePath) throws IOException {
```

Comment:
nit: no used function `getAllFiles`, could we remove it？ or Am I missing something？